File: linkList/5.linkedListCycleII/go/linkedListCycleII.go
package _go

type ListNode struct {
	Val  int
	Next *ListNode
}

func detectCycle(head *ListNode) *ListNode {

	// 快慢指针
	slow := head
	quick := head
	for {
		if quick == nil || quick.Next == nil {
			return nil
		}
		slow = slow.Next
		quick = quick.Next.Next
		if quick == slow {
			break
		}
	}
	// 快指针返回头节点,变成慢指针
	quick = head
	for quick != slow {
		slow = slow.Next
		quick = quick.Next
	}
	return quick
}

File: linkList/5.linkedListCycleII/py/linkedListCycleII.py
from typing import Optional


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None
class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow = head
        fast = head
        while True:
            print(1)
            if not(fast and fast.next):
                return None
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                break
        # 快指针回到原点，变成慢指针
        fast = head
        while fast != slow:
            fast = fast.next
            slow = slow.next
        return fast